File: Typescript/src/clope/MathCache.ts
export default class MathCache {
    private repulsion: number;
    private precalculatedWPowR: number[];
    constructor(repulsion: number) {
        this.repulsion = repulsion;
        this.precalculatedWPowR = [];
    }

    /**
     * Gradient function for profit counting
     * @param S Square of cluster
     * @param N Number of objects in cluster
     * @param width width of cluster
     */
    public Grad(S: number, N: number, width: number): number {
        return S * N / this.GetWPowR(width);
    }

    private GetWPowR(w: number) {
        if (w >= this.precalculatedWPowR.length) { this.Prepare(w); }
        return this.precalculatedWPowR[w];
    }

    private Prepare(w: number): void {
        for (let i = this.precalculatedWPowR.length; i <= w; i++) {
            this.precalculatedWPowR.push(Math.pow(i, this.repulsion));
        }
    }
}

File: c#/Clope/clope/MathCache.cs
using System.Collections.Generic;

namespace Clope.clope
{
    public class MathCache
    {
        private double Repulsion { get; }
        private readonly List<double> _preCalculatedWpowR;
        public MathCache(double repulsion)
        {
            Repulsion = repulsion;
            _preCalculatedWpowR = new List<double>();
        }

        public double Grad(int S, int N, int width)
        {
            return S * N / GetWPowR(width);
        }

        private double GetWPowR(int w)
        {
            if (w >= _preCalculatedWpowR.Count) Prepare(w);
            return _preCalculatedWpowR[w];
        }

        private void Prepare(int w)
        {
            for (var i = _preCalculatedWpowR.Count; i <= w; i++)
                _preCalculatedWpowR.Add(System.Math.Pow(i, Repulsion));
        }


    }
}